### PR TITLE
Plugin: Use bundled files for enqueuing global styles

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
Regression after https://github.com/WordPress/gutenberg/pull/50079#issuecomment-1533905463

PR adds a script loader not tied to the specific WP version to override global styles enqueued by core. The new one will use the `WP_Theme_JSON` classes bundled with the plugin.

## Why?
Bundled `WP_Theme_JSON` files should provide some of the settings for global styles.

## Testing Instructions
https://github.com/WordPress/gutenberg/pull/50079#issuecomment-1535531248

### CODE REVIEW

**Comment by Mamaduka** [comment]
Hey, @oandregal

What do you think about these "permanent" overrides for settings that need to use bundled `WP_Theme_JSON_*Gutenberg` classes?

**Review by tellthemachines — COMMENTED** [review]
Thanks for getting this started! It not working for me yet with the testing steps I outlined in the other PR, but I think this is a good approach in general, at least while global styles and design tools are still under active development.

**Comment by Mamaduka** [comment]
@tellthemachines, I think I got all the parts ported into the new system. Overriding the `styles` for block editor settings was the missing piece - eec751aa9a9c299330917a8cc824c3c6ef169ebb.

File: lib/block-editor-settings.php
Comment (Mamaduka) [inline lib/block-editor-settings.php]: I renamed the method in the `compat/wp-6.2` directory with a guard clause, so we don't have to maintain a separate copy forever.

**Comment by oandregal** [comment]
> What do you think about these "permanent" overrides for settings that need to use bundled WP_Theme_JSON_*Gutenberg classes?

Yeah, this is annoying. As long as the underlying classes receive changes, the public API needs to point at them in the plugin.

I'd think it's best to make them "evergreen" (so they'd sit in `lib/`) as the private classes do. I wouldn't be against making them part of `lib/compat/6.X`, though it seems that this would just "move the problem to the future": once the minimum WP version for Gutenberg is 6.2 they'll be removed, so we'll run into the same issue because the private classes will still receive updates. Does that sound sensible?

**Comment by Mamaduka** [comment]
> I wouldn't be against making them part of `lib/compat/6.X`, though it seems that this would just "move the problem to the future"

I agree. Keeping this code in `compat/wordpress-` creates an "opportunity" for future regressions.

One more thing. Should I also move `_gutenberg_clean_theme_json_caches` into the "evergreen" files, or is there a plan to match the cache keys used by the plugin to WP?

**Comment by oandregal** [comment]
> One more thing. Should I also move _gutenberg_clean_theme_json_caches into the "evergreen" files, or is there a plan to match the cache keys used by the plugin to WP?

No, they need to be different from core.

Would you be up to prepare small PRs that migrate each method separately? Otherwise, it's a big PR with a lot of changes that are difficult to match to each other. If it's just moving methods around, things are easy to spot.

**Comment by Mamaduka** [comment]
> Would you be up to prepare small PRs that migrate each method separately?

I would, but all the changes in this PR are needed to fix the regression reported by @tellthemachines - https://github.com/WordPress/gutenberg/pull/50079#issuecomment-1533905463. 

I've not touched the methods; I just moved them. The changes that could use a detailed review are concentrated in [`lib/block-editor-settings.php`](https://github.com/WordPress/gutenberg/pull/50310/files#diff-194fa91cbbc3043754b8f48d4ab6964bf37abdef44d11b07b24c92d8a651833a) and [`lib/script-loader.php`](https://github.com/WordPress/gutenberg/pull/50310/files#diff-18907ef47b0dcd4a0c5d378acf15f3d41b1ac434714569f3f0004c0a2a5f59f8) files.

File: lib/compat/wordpress-6.2/block-editor-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/block-editor-settings.php]: Given that we have a `gutenberg_get_block_editor_settings` in `lib/block-editor-settings.php`, why do we need `lib/compat/wordpress-6.2/block-editor-settings.php` as well?

File: lib/block-editor-settings.php
Comment (oandregal) [reply lib/block-editor-settings.php]: :thinking: Isn't the point of this PR to make sure we use the gutenberg method that points to the gutenberg class?

File: lib/compat/wordpress-6.2/script-loader.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/script-loader.php]: :thinking: same feedback as above. Isn't the point of this PR to make sure we use the gutenberg methods that point to the gutenberg classes? This is a behavior change: as soon as the core function exists, it'll be used instead of gutenberg.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php]: I was expecting that the only change here would be to move some methods from this file to `lib/get-global-styles-and-settings.php`.

As far as I've checked, we should move `gutenberg_get_global_styles_custom_css`, `gutenberg_get_global_stylesheet`, `gutenberg_get_global_settings`, `_gutenberg_clean_theme_json_caches`, `_gutenberg_add_non_persistent_theme_json_cache_group`.

File: lib/compat/wordpress-6.2/block-editor-settings.php
Comment (Mamaduka) [reply lib/compat/wordpress-6.2/block-editor-settings.php]: Good catch. I'll remove that.

File: lib/script-loader.php
Comment (oandregal) [inline lib/script-loader.php:9]: So, besides moving things around, we are also removing core callbacks and hooking only gutenberg's. That's great. We'll probably need to do it for a few other things: `wp_enqueue_global_styles_custom_css` and `wp_enqueue_global_styles_css_custom_properties`.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (Mamaduka) [reply lib/compat/wordpress-6.2/get-global-styles-and-settings.php]: I moved everything besides `_gutenberg_add_non_persistent_theme_json_cache_group`. The cache group would be the same, so there's no need to keep that function after 6.2.

**Comment by Mamaduka** [comment]
@oandregal, I addressed the feedback, and PR is ready for another review.

**Comment by Mamaduka** [comment]
Hi, @tellthemachines

Did you get a chance to test the fix?

**Comment by tellthemachines** [comment]
@Mamaduka I won't be able to do a full code review this week, but I've just tested the latest state of this branch and it's now working correctly. I checked the spacing styles within flow/constrained layouts and also the experimental grid styles, and both are as expected ✅

**Comment by Mamaduka** [comment]
> I won't be able to do a full code review this week,

Not a problem. I just want to be sure that PR fixes the regression. Thanks for testing!

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for addressing this @Mamaduka 👍 

It's always a pleasure to find a bug already has a fix up for review!

It turns out https://github.com/WordPress/gutenberg/pull/50079 also caused the custom [selectors API  to be missed](https://github.com/WordPress/gutenberg/issues/50416). This PR fixes that issue.

The code changes LGTM but it might be best to wait on a final review from @tellthemachines to make sure everything has been covered from their end.

**Comment by Mamaduka** [comment]
Thanks for testing, @aaronrobertshaw!

@oandregal, do you mind giving this PR another review?

File: lib/block-editor-settings.php
Comment (oandregal) [inline lib/block-editor-settings.php:17]: Documenting what I learn: this is mostly a copy-over of `get_block_editor_settings` but using the `gutenberg_*` functions instead of the `wp_*` ones.

File: lib/compat/wordpress-6.2/block-editor-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/block-editor-settings.php:15]: Documenting what I learn: this has been moved to `lib/block-editor-settings.php`, which includes this code (and much more).

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:67]: This has been moved to `lib/global-styles-and-settings.php` verbatim.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:104]: This has been moved to `lib/global-styles-and-settings.php` verbatim.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:188]: This has been moved to `lib/global-styles-and-settings.php` verbatim.

File: lib/compat/wordpress-6.2/get-global-styles-and-settings.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/get-global-styles-and-settings.php:223]: This has been moved to `lib/global-styles-and-settings.php` verbatim.

File: lib/compat/wordpress-6.2/default-filters.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/default-filters.php:20]: This has been moved to `lib/global-styles-and-settings.php` verbatim. The comment no longer makes sense as core already uses [the proper function as a callback](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/default-filters.php#L356).

File: lib/compat/wordpress-6.2/script-loader.php
Comment (oandregal) [inline lib/compat/wordpress-6.2/script-loader.php:161]: The function and the filter has been moved to `lib/script-loader.php` verbatim.

File: lib/global-styles-and-settings.php
Comment (oandregal) [inline lib/global-styles-and-settings.php:170]: This brings back this function, that had been removed in https://github.com/WordPress/gutenberg/pull/50079

File: lib/script-loader.php
Comment (oandregal) [reply lib/script-loader.php:9]: The previous comment is incorrect: this is actually bringing back something we removed in https://github.com/WordPress/gutenberg/pull/50079

File: lib/script-loader.php
Comment (oandregal) [inline lib/script-loader.php:19]: This brings this function back after having being removed at https://github.com/WordPress/gutenberg/pull/50079

File: lib/script-loader.php
Comment (oandregal) [inline lib/script-loader.php:88]: This brings back a function that I haven't seen at https://github.com/WordPress/gutenberg/pull/50079 though it still makes sense to me.

**Review by oandregal — APPROVED** [review]
I haven't tested this solves the issues brought by others. The code changes make sense to me: either move things around or add back things (added comments for each function trying to document what changes).

**Comment by Mamaduka** [comment]
Thanks, @oandregal!

I’ll merge this to fix the regressions. Happy to follow up in case I missed something.

**Comment by Mamaduka** [comment]
@fabiankaegy, the #50079 will ship with Gutenberg 15.8; I think it makes sense to cherry-pick this PR and avoid regressions in the release.

**Comment by fabiankaegy** [comment]
I just cherry-picked this PR to the release/15.8 branch to get it included in the next release: f7d2fbd216

**Comment by oandregal** [comment]
I've got a couple of follow-ups at https://github.com/WordPress/gutenberg/pull/50596 and https://github.com/WordPress/gutenberg/pull/50597.

File: lib/block-editor-settings.php
Comment (ajlende) [inline lib/block-editor-settings.php:75]: It seems this broke duotone filters which rely on updating `$settings['styles']` in the same hook. Is there a reason that this hook needs to run at `PHP_INT_MAX` priority?

Comment (Mamaduka) [reply lib/block-editor-settings.php:75]: > Is there a reason that this hook needs to run at PHP_INT_MAX priority?

That's what the plugin always used for [block setting overrides](https://github.com/WordPress/gutenberg/blob/1664d5e9708d6085c003aaef2d2feb9343f0e0d6/lib/compat/wordpress-6.2/block-editor-settings.php#L31). I guess to prevent other plugins from overriding new feature settings.

I personally don't like the use of `PHP_INT_MAX`; it prevents extensibility.

Can you point me to where the Duotone filters are located? Happy to help with the fix.

Comment (ajlende) [reply lib/block-editor-settings.php:75]: > That's what the plugin always used for [block setting overrides](https://github.com/WordPress/gutenberg/blob/1664d5e9708d6085c003aaef2d2feb9343f0e0d6/lib/compat/wordpress-6.2/block-editor-settings.php#L31). 

Ah, it just appended to the styles rather than replacing them before. 

> Can you point me to where the Duotone filters are located?

The duotone hook is here https://github.com/WordPress/gutenberg/blob/b05777639618d8c3b79451c1570b5c0a881910f5/lib/block-supports/duotone.php#L54

> Happy to help with the fix.

Thanks! I'll probably push up a PR sometime today and add you as a reviewer. It sounds like it should be safe to lower the priority to allow for the extensibility that now comes from within the plugin.

Comment (Mamaduka) [reply lib/block-editor-settings.php:75]: @oandregal, any thoughts on lowering the priority for the plugin overrides?

Comment (ajlende) [reply lib/block-editor-settings.php:75]: Opened https://github.com/WordPress/gutenberg/pull/50760. I found and fixed a few other things when I was digging through all of the code that used the same hook.

**Pull request merged**
